File: src/harness/security/SensitiveDetector.ts
/**
 * 统一敏感信息检测器 — 全局唯一实现
 *
 * 消除 ConstraintsService / VerificationService / IndependentEvaluationService
 * 三处重复的敏感信息检测逻辑，改为统一调用此模块。
 *
 * 设计原则：
 * - 只做检测，不做拦截决策（拦截由调用方决定）
 * - 返回结构化结果，支持不同场景的差异化处理
 * - 检测模式集中维护，避免散落各处
 */

/** 单条违规记录 */
export interface SensitiveViolation {
  /** 违规类型名称 */
  name: string;
  /** 风险等级 */
  risk: 'low' | 'medium' | 'high' | 'critical';
  /** 匹配到的原始文本（可选，用于日志） */
  matchedText?: string;
}

/** 检测结果 */
export interface SensitiveCheckResult {
  /** 是否安全（无违规） */
  safe: boolean;
  /** 综合风险等级 */
  riskLevel: 'none' | 'low' | 'medium' | 'high' | 'critical';
  /** 违规列表 */
  violations: SensitiveViolation[];
  /** 脱敏后的文本（仅当有违规时生成） */
  sanitizedOutput?: string;
}

/** 检测场景 — 不同场景使用不同的模式子集 */
export type SensitiveCheckScene =
  | 'output'      // LLM 输出检测：全量模式
  | 'storage'     // 存储前检测：侧重凭证和密钥
  | 'command';    // 命令检测：侧重危险操作

/** 敏感信息模式定义 */
interface SensitivePattern {
  pattern: RegExp;
  name: string;
  risk: SensitiveViolation['risk'];
  /** 适用场景，不指定则全场景适用 */
  scenes?: SensitiveCheckScene[];
}

/**
 * 统一敏感信息模式库
 *
 * 维护规则：
 * - 新增模式只在此处添加，不要在其他文件重复定义
 * - 每个模式标注 risk 和适用 scenes
 * - 正则使用 g 标志以支持 test + replace
 */
const SENSITIVE_PATTERNS: SensitivePattern[] = [
  // ===== 金融类 =====
  { pattern: /\b\d{16,19}\b/g, name: '银行卡号', risk: 'high' },
  { pattern: /\b\d{6}\d{4}\d{2}\d{2}\d{4}\b/g, name: '身份证号', risk: 'high' },
  { pattern: /\b\d{4}[/\-]?\d{2}[/\-]?\d{2}\b/g, name: '银行卡有效期', risk: 'medium' },
  { pattern: /\bCVV[:\s]*\d{3,4}\b/gi, name: 'CVV码', risk: 'critical' },
  { pattern: /\b\d{17}[\dXx]\b/g, name: '身份证号(18位)', risk: 'high' },

  // ===== 认证凭据 =====
  { pattern: /(?:password|密码|pwd|passwd)\s*[:=]\s*\S+/gi, name: '密码泄露', risk: 'critical' },
  { pattern: /(?:secret|密钥|api[_-]?key|access[_-]?token|secret[_-]?key)\s*[:=]\s*['"]?[a-zA-Z0-9]{8,}/gi, name: '密钥/Token泄露', risk: 'critical', scenes: ['output', 'storage'] },
  { pattern: /(?:bearer|basic)\s+\S+/gi, name: '认证头泄露', risk: 'high', scenes: ['output'] },
  { pattern: /\b(?:sk-|api_)[a-zA-Z0-9]{20,}/g, name: 'API密钥', risk: 'critical' },
  { pattern: /\bAKIA[A-Z0-9]{16}\b/g, name: 'AWS访问密钥', risk: 'critical' },
  { pattern: /\bghp_[a-zA-Z0-9]{36}\b/g, name: 'GitHub令牌', risk: 'critical' },
  { pattern: /\bgho_[a-zA-Z0-9]{36}\b/g, name: 'GitHub OAuth令牌', risk: 'critical' },
  { pattern: /\bxox[baprs]-[a-zA-Z0-9]{10,}/g, name: 'Slack令牌', risk: 'critical' },

  // ===== 存储专用：更严格的凭证检测 =====
  { pattern: /\bsk-[a-zA-Z0-9]{8,}/g, name: 'API密钥', risk: 'critical', scenes: ['storage'] },
  { pattern: /\bAKIA[A-Z0-9]{16}\b/g, name: 'AWS密钥', risk: 'critical', scenes: ['storage'] },
  { pattern: /\bghp_[a-zA-Z0-9]{36}\b/g, name: 'GitHub令牌', risk: 'critical', scenes: ['storage'] },
  { pattern: /(?:api[_-]?key|apikey|access[_-]?token|secret[_-]?key)\s*[:=]\s*['"]?[a-zA-Z0-9]{8,}/i, name: '密钥凭证', risk: 'critical', scenes: ['storage'] },
  { pattern: /密钥|密码|口令|私钥|secret|credential/i, name: '敏感凭证关键词', risk: 'high', scenes: ['storage'] },

  // ===== 通信联系方式 =====
  { pattern: /\b1[3-9]\d{9}\b/g, name: '手机号码', risk: 'medium' },
  { pattern: /\+86[-\s]?1[3-9]\d{9}\b/g, name: '中国手机号码', risk: 'medium' },
  { pattern: /\b0\d{2,3}[-\s]?\d{7,8}\b/g, name: '固话号码', risk: 'medium' },
  { pattern: /\b[A-Za-z0-9._%+-]+@[A-Za-z0-9.-]+\.[A-Z|a-z]{2,}\b/g, name: '邮箱地址', risk: 'medium' },

  // ===== 网络地址 =====
  { pattern: /\b\d{1,3}\.\d{1,3}\.\d{1,3}\.\d{1,3}\b/g, name: 'IPv4地址', risk: 'low' },
  { pattern: /\b(?:[0-9a-fA-F]{1,4}:){7}[0-9a-fA-F]{1,4}\b/g, name: 'IPv6地址', risk: 'low' },
  { pattern: /::(?:[0-9a-fA-F]{1,4}:){0,5}[0-9a-fA-F]{1,4}\b/gi, name: 'IPv6地址(压缩)', risk: 'low' },
  { pattern: /\b[0-9a-fA-F]{2}(?:[:-][0-9a-fA-F]{2}){5}\b/g, name: 'MAC地址', risk: 'low' },

  // ===== 其他个人信息 =====
  { pattern: /\b[A-Z]\d{8,9}\b/g, name: '护照号', risk: 'high' },
  { pattern: /(?:病历|处方|诊断)[:：]\S+/gi, name: '医疗信息', risk: 'high' },
];

/**
 * 危险命令模式库
 */
const DANGEROUS_COMMAND_PATTERNS: RegExp[] = [
  /\brm\s+-rf\s+\//,
  /\bdel\s+\/f\s+\/q\s+/i,
  /\bformat\s+[A-Za-z]:/i,
  /\bshutdown\b/,
  /\bdrop\s+table\b/i,
  /\bdrop\s+database\b/i,
  /\btruncate\b.*\btable\b/i,
  /\b--\s*;\s*drop\b/i,
];

/**
 * 统一敏感信息检测入口
 *
 * @param text - 待检测文本
 * @param scene - 检测场景，不同场景使用不同模式子集
 * @returns 检测结果
 */
export function checkSensitiveInfo(
  text: string,
  scene: SensitiveCheckScene = 'output'
): SensitiveCheckResult {
  const violations: SensitiveViolation[] = [];

  const applicablePatterns = SENSITIVE_PATTERNS.filter(
    (p) => !p.scenes || p.scenes.includes(scene)
  );

  for (const { pattern, name, risk } of applicablePatterns) {
    // 每次检测需要新建 RegExp 实例，避免 lastIndex 残留
    const regex = new RegExp(pattern.source, pattern.flags);
    if (regex.test(text)) {
      violations.push({ name, risk });
    }
  }

  const hasViolations = violations.length > 0;
  const hasCritical = violations.some((v) => v.risk === 'critical');
  const hasHigh = violations.some((v) => v.risk === 'high');

  let sanitizedOutput: string | undefined;
  if (hasViolations) {
    sanitizedOutput = sanitizeText(text);
  }

  return {
    safe: !hasViolations,
    riskLevel: hasCritical
      ? 'critical'
      : hasHigh
        ? 'high'
        : violations.length > 0
          ? 'medium'
          : 'none',
    violations,
    sanitizedOutput,
  };
}

/**
 * 检测危险命令
 *
 * @param command - 待检测的命令字符串
 * @returns 是否为危险命令及原因
 */
export function checkDangerousCommand(
  command: string
): { dangerous: boolean; reason?: string } {
  for (const pattern of DANGEROUS_COMMAND_PATTERNS) {
    if (pattern.test(command)) {
      return {
        dangerous: true,
        reason: `检测到危险命令模式: ${command.substring(0, 50)}`,
      };
    }
  }
  return { dangerous: false };
}

/**
 * 文本脱敏处理
 *
 * 将敏感信息替换为脱敏标记
 */
export function sanitizeText(text: string): string {
  return text
    // API密钥（优先处理，避免被后续规则误匹配）
    .replace(/(?:sk-|api_)[a-zA-Z0-9]{20,}/gi, '[API密钥-已脱敏]')
    // AWS密钥
    .replace(/\bAKIA[A-Z0-9]{16}\b/g, '[AWS密钥-已脱敏]')
    // GitHub令牌
    .replace(/\bghp_[a-zA-Z0-9]{36}\b/g, '[GitHub令牌-已脱敏]')
    .replace(/\bgho_[a-zA-Z0-9]{36}\b/g, '[GitHub OAuth-已脱敏]')
    // Slack令牌
    .replace(/\bxox[baprs]-[a-zA-Z0-9]{10,}/g, '[Slack令牌-已脱敏]')
    // 银行卡号
    .replace(/\b\d{16,19}\b/g, '[银行卡-已脱敏]')
    // 身份证号
    .replace(/\b\d{6}\d{4}\d{2}\d{2}\d{4}\b/g, '[身份证-已脱敏]')
    .replace(/\b\d{17}[\dXx]\b/g, '[身份证-已脱敏]')
    // 手机号
    .replace(/\b1[3-9]\d{9}\b/g, '[手机号-已脱敏]')
    .replace(/\+86[-\s]?1[3-9]\d{9}\b/g, '[手机号-已脱敏]')
    // 邮箱
    .replace(/\b[A-Za-z0-9._%+-]+@[A-Za-z0-9.-]+\.[A-Z|a-z]{2,}\b/g, '[邮箱-已脱敏]')
    // IPv4
    .replace(/\b\d{1,3}\.\d{1,3}\.\d{1,3}\.\d{1,3}\b/g, '[IP-已脱敏]')
    // IPv6
    .replace(/\b(?:[0-9a-fA-F]{1,4}:){7}[0-9a-fA-F]{1,4}\b/gi, '[IPv6-已脱敏]')
    .replace(/::(?:[0-9a-fA-F]{1,4}:){0,5}[0-9a-fA-F]{1,4}\b/gi, '[IPv6-已脱敏]')
    // 密码/密钥
    .replace(/((?:password|密码|pwd|passwd|secret|密钥|api[_-]?key|token)\s*[:=]\s*)\S+/gi, '$1[已脱敏]')
    // 认证头
    .replace(/(?:bearer|basic)\s+\S+/gi, '[认证头-已脱敏]')
    // MAC地址
    .replace(/\b[0-9a-fA-F]{2}(?:[:-][0-9a-fA-F]{2}){5}\b/g, '[MAC-已脱敏]')
    // 护照号
    .replace(/\b[A-Z]\d{8,9}\b/g, '[护照号-已脱敏]');
}

File: python/agent/security/sensitive_detector.py
from __future__ import annotations

import re
from dataclasses import dataclass, field
from enum import Enum
from typing import Any

from agent.core.types import RiskLevel


class CheckScene(str, Enum):
    """检查场景枚举。

    Attributes:
        OUTPUT: 输出内容检查。
        STORAGE: 存储内容检查。
        COMMAND: 命令执行检查。
    """

    OUTPUT = "output"
    STORAGE = "storage"
    COMMAND = "command"


@dataclass
class SensitiveViolation:
    """敏感信息违规记录。

    Attributes:
        name: 违规类型名称。
        risk: 风险等级。
        matched_text: 匹配到的敏感文本。
    """

    name: str
    risk: RiskLevel
    matched_text: str | None = None


@dataclass
class SensitiveCheckResult:
    """敏感信息检查结果。

    Attributes:
        safe: 是否安全。
        risk_level: 整体风险等级。
        violations: 违规列表。
        sanitized_output: 脱敏后的输出。
    """

    safe: bool
    risk_level: RiskLevel
    violations: list[SensitiveViolation] = field(default_factory=list)
    sanitized_output: str | None = None


@dataclass
class DangerousCommandResult:
    """危险命令检查结果。

    Attributes:
        dangerous: 是否危险。
        reason: 危险原因。
    """

    dangerous: bool
    reason: str | None = None


@dataclass
class SensitivePattern:
    """敏感信息匹配模式。

    Attributes:
        pattern: 正则表达式。
        name: 模式名称。
        risk: 风险等级。
        scenes: 适用的检查场景。
    """

    pattern: re.Pattern[str]
    name: str
    risk: RiskLevel
    scenes: list[CheckScene] | None = None


_SENSITIVE_PATTERNS: list[SensitivePattern] = [
    SensitivePattern(re.compile(r"\b\d{16,19}\b"), "银行卡号", RiskLevel.HIGH),
    SensitivePattern(re.compile(r"\b\d{6}\d{4}\d{2}\d{2}\d{4}\b"), "身份证号", RiskLevel.HIGH),
    SensitivePattern(re.compile(r"\b\d{4}[/\-]?\d{2}[/\-]?\d{2}\b"), "银行卡有效期", RiskLevel.MEDIUM),
    SensitivePattern(re.compile(r"\bCVV[:\s]*\d{3,4}\b", re.IGNORECASE), "CVV码", RiskLevel.CRITICAL),
    SensitivePattern(re.compile(r"\b\d{17}[\dXx]\b"), "身份证号(18位)", RiskLevel.HIGH),
    SensitivePattern(re.compile(r"(?:password|密码|pwd|passwd)\s*(?:[:=]|是|is|was|are)\s*\S+", re.IGNORECASE), "密码泄露", RiskLevel.CRITICAL),
    SensitivePattern(re.compile(r"(?:secret|密钥|api[_-]?key|access[_-]?token|secret[_-]?key)\s*(?:[:=]|是|is|was|are)\s*['\"]?[a-zA-Z0-9\-]{8,}", re.IGNORECASE), "密钥/Token泄露", RiskLevel.CRITICAL, [CheckScene.OUTPUT, CheckScene.STORAGE]),
    SensitivePattern(re.compile(r"(?:bearer|basic)\s+\S+", re.IGNORECASE), "认证头泄露", RiskLevel.HIGH, [CheckScene.OUTPUT]),
    SensitivePattern(re.compile(r"\b(?:sk-|api_)[a-zA-Z0-9]{20,}"), "API密钥", RiskLevel.CRITICAL),
    SensitivePattern(re.compile(r"\bAKIA[A-Z0-9]{16}\b"), "AWS访问密钥", RiskLevel.CRITICAL),
    SensitivePattern(re.compile(r"\bghp_[a-zA-Z0-9]{36}\b"), "GitHub令牌", RiskLevel.CRITICAL),
    SensitivePattern(re.compile(r"\bgho_[a-zA-Z0-9]{36}\b"), "GitHub OAuth令牌", RiskLevel.CRITICAL),
    SensitivePattern(re.compile(r"\bxox[baprs]-[a-zA-Z0-9]{10,}"), "Slack令牌", RiskLevel.CRITICAL),
    SensitivePattern(re.compile(r"\bsk-[a-zA-Z0-9]{8,}"), "API密钥", RiskLevel.CRITICAL, [CheckScene.STORAGE]),
    SensitivePattern(re.compile(r"\bAKIA[A-Z0-9]{16}\b"), "AWS密钥", RiskLevel.CRITICAL, [CheckScene.STORAGE]),
    SensitivePattern(re.compile(r"\bghp_[a-zA-Z0-9]{36}\b"), "GitHub令牌", RiskLevel.CRITICAL, [CheckScene.STORAGE]),
    SensitivePattern(re.compile(r"(?:api[_-]?key|apikey|access[_-]?token|secret[_-]?key)\s*[:=]\s*['\"]?[a-zA-Z0-9]{8,}", re.IGNORECASE), "密钥凭证", RiskLevel.CRITICAL, [CheckScene.STORAGE]),
    SensitivePattern(re.compile(r"密钥|密码|口令|私钥|secret|credential", re.IGNORECASE), "敏感凭证关键词", RiskLevel.HIGH, [CheckScene.STORAGE]),
    SensitivePattern(re.compile(r"(?<!\d)1[3-9]\d{9}(?!\d)"), "手机号码", RiskLevel.MEDIUM),
    SensitivePattern(re.compile(r"\+86[-\s]?1[3-9]\d{9}(?!\d)"), "中国手机号码", RiskLevel.MEDIUM),
    SensitivePattern(re.compile(r"\b0\d{2,3}[-\s]?\d{7,8}\b"), "固话号码", RiskLevel.MEDIUM),
    SensitivePattern(re.compile(r"\b[A-Za-z0-9._%+\-]+@[A-Za-z0-9.\-]+\.[A-Za-z]{2,}\b"), "邮箱地址", RiskLevel.MEDIUM),
    SensitivePattern(re.compile(r"\b\d{1,3}\.\d{1,3}\.\d{1,3}\.\d{1,3}\b"), "IPv4地址", RiskLevel.LOW),
    SensitivePattern(re.compile(r"\b(?:[0-9a-fA-F]{1,4}:){7}[0-9a-fA-F]{1,4}\b"), "IPv6地址", RiskLevel.LOW),
    SensitivePattern(re.compile(r"::(?:[0-9a-fA-F]{1,4}:){0,5}[0-9a-fA-F]{1,4}\b", re.IGNORECASE), "IPv6地址(压缩)", RiskLevel.LOW),
    SensitivePattern(re.compile(r"\b[0-9a-fA-F]{2}(?:[:-][0-9a-fA-F]{2}){5}\b"), "MAC地址", RiskLevel.LOW),
    SensitivePattern(re.compile(r"\b[A-Z]\d{8,9}\b"), "护照号", RiskLevel.HIGH),
    SensitivePattern(re.compile(r"(?:病历|处方|诊断)[:：]\S+", re.IGNORECASE), "医疗信息", RiskLevel.HIGH),
]

_DANGEROUS_COMMAND_PATTERNS: list[re.Pattern[str]] = [
    re.compile(r"\brm\s+-rf\s+/"),
    re.compile(r"\bdel\s+/f\s+/q\s+", re.IGNORECASE),
    re.compile(r"\bdel\s+/f\s+", re.IGNORECASE),
    re.compile(r"\bformat\s+[A-Za-z]:", re.IGNORECASE),
    re.compile(r"\bshutdown\b"),
    re.compile(r"\bdrop\s+table\b", re.IGNORECASE),
    re.compile(r"\bdrop\s+database\b", re.IGNORECASE),
    re.compile(r"\btruncate\b.*\btable\b", re.IGNORECASE),
    re.compile(r"\b--\s*;\s*drop\b", re.IGNORECASE),
]


def check_sensitive_info(
    text: str,
    scene: CheckScene = CheckScene.OUTPUT,
) -> SensitiveCheckResult:
    """检查文本中的敏感信息。

    扫描文本中的敏感信息（密钥、密码、身份证号、手机号等），
    返回风险等级和违规列表。High以上风险文本将自动脱敏。

    Args:
        text: 待检查的文本。
        scene: 检查场景（OUTPUT/STORAGE/COMMAND）。

    Returns:
        SensitiveCheckResult: 包含安全状态、风险等级、违规列表和脱敏文本。
    """
    violations: list[SensitiveViolation] = []

    for sp in _SENSITIVE_PATTERNS:
        if sp.scenes and scene not in sp.scenes:
            continue
        regex = re.compile(sp.pattern.pattern, sp.pattern.flags)
        if regex.search(text):
            violations.append(SensitiveViolation(name=sp.name, risk=sp.risk))

    has_violations = len(violations) > 0
    has_critical = any(v.risk == RiskLevel.CRITICAL for v in violations)
    has_high = any(v.risk == RiskLevel.HIGH for v in violations)

    sanitized_output: str | None = None
    if has_violations:
        sanitized_output = sanitize_text(text)

    if has_critical:
        risk_level = RiskLevel.CRITICAL
    elif has_high:
        risk_level = RiskLevel.HIGH
    elif has_violations:
        risk_level = RiskLevel.MEDIUM
    else:
        risk_level = RiskLevel.NONE

    return SensitiveCheckResult(
        safe=not has_violations,
        risk_level=risk_level,
        violations=violations,
        sanitized_output=sanitized_output,
    )


def check_dangerous_command(command: str) -> DangerousCommandResult:
    for pattern in _DANGEROUS_COMMAND_PATTERNS:
        if pattern.search(command):
            return DangerousCommandResult(
                dangerous=True,
                reason=f"检测到危险命令模式: {command[:50]}",
            )
    return DangerousCommandResult(dangerous=False)


def sanitize_text(text: str) -> str:
    result = text
    result = re.sub(r"(?:sk-|api_)[a-zA-Z0-9]{20,}", "[API密钥-已脱敏]", result, flags=re.IGNORECASE)
    result = re.sub(r"\bAKIA[A-Z0-9]{16}\b", "[AWS密钥-已脱敏]", result)
    result = re.sub(r"\bghp_[a-zA-Z0-9]{36}\b", "[GitHub令牌-已脱敏]", result)
    result = re.sub(r"\bgho_[a-zA-Z0-9]{36}\b", "[GitHub OAuth-已脱敏]", result)
    result = re.sub(r"\bxox[baprs]-[a-zA-Z0-9]{10,}", "[Slack令牌-已脱敏]", result)
    result = re.sub(r"\b\d{16,19}\b", "[银行卡-已脱敏]", result)
    result = re.sub(r"\b\d{6}\d{4}\d{2}\d{2}\d{4}\b", "[身份证-已脱敏]", result)
    result = re.sub(r"\b\d{17}[\dXx]\b", "[身份证-已脱敏]", result)
    result = re.sub(r"(?<!\d)1[3-9]\d{9}(?!\d)", "[手机号-已脱敏]", result)
    result = re.sub(r"\+86[-\s]?1[3-9]\d{9}(?!\d)", "[手机号-已脱敏]", result)
    result = re.sub(r"\b[A-Za-z0-9._%+\-]+@[A-Za-z0-9.\-]+\.[A-Za-z]{2,}\b", "[邮箱-已脱敏]", result)
    result = re.sub(r"\b\d{1,3}\.\d{1,3}\.\d{1,3}\.\d{1,3}\b", "[IP-已脱敏]", result)
    result = re.sub(r"\b(?:[0-9a-fA-F]{1,4}:){7}[0-9a-fA-F]{1,4}\b", "[IPv6-已脱敏]", result, flags=re.IGNORECASE)
    result = re.sub(r"::(?:[0-9a-fA-F]{1,4}:){0,5}[0-9a-fA-F]{1,4}\b", "[IPv6-已脱敏]", result, flags=re.IGNORECASE)
    result = re.sub(r"((?:password|密码|pwd|passwd|secret|密钥|api[_-]?key|token)\s*(?:[:=]|是|is|was|are)\s*)\S+", r"\1[已脱敏]", result, flags=re.IGNORECASE)
    result = re.sub(r"(?:bearer|basic)\s+\S+", "[认证头-已脱敏]", result, flags=re.IGNORECASE)
    result = re.sub(r"\b[0-9a-fA-F]{2}(?:[:-][0-9a-fA-F]{2}){5}\b", "[MAC-已脱敏]", result)
    result = re.sub(r"\b[A-Z]\d{8,9}\b", "[护照号-已脱敏]", result)
    return result
